pkg/controller/instance/instance_controller.go

```diff
@@ -30,12 +30,27 @@ var log = logf.Log.WithName("controller_harbor")
 // Add creates a new Instance Controller and adds it to the Manager. The Manager will set fields on the Controller
 // and Start it when the Manager is Started.
 func Add(mgr manager.Manager) error {
-	return add(mgr, newReconciler(mgr))
+	// This function is used to dynamically generate a helmclient
+	// and is passed as a field value to the ReconcileInstance struct.
+	f := func(repoCache, repoConfig, namespace string) (*helmclient.Client, error) {
+		opts := &helmclient.RestConfClientOptions{
+			Options: &helmclient.Options{
+				Namespace:        namespace,
+				RepositoryCache:  repoCache,
+				RepositoryConfig: repoConfig,
+			},
+			RestConfig: mgr.GetConfig(),
+		}
+
+		return helmclient.NewClientFromRestConf(opts)
+	}
+
+	return add(mgr, newReconciler(mgr, f))
 }
 
 // newReconciler returns a new reconcile.Reconciler
-func newReconciler(mgr manager.Manager) reconcile.Reconciler {
-	return &ReconcileInstance{client: mgr.GetClient(), scheme: mgr.GetScheme(), config: mgr.GetConfig()}
+func newReconciler(mgr manager.Manager, f func(repoCache, repoConfig, namespace string)(*helmclient.Client, error)) reconcile.Reconciler {
```

Comment:
Maybe extract the function type into a named type?

```go
type helmClientFactory func(repoCache, repoConfig, namespace string)(*helmclient.Client, error)
```

Reply:
Good one! Changed it.